libsolidity/ast/AST.cpp

```diff
@@ -314,11 +302,7 @@ ContractDefinition::ContractKind FunctionDefinition::inContractKind() const
 
 CallableDeclarationAnnotation& CallableDeclaration::annotation() const
 {
-	solAssert(
-		m_annotation,
-		"CallableDeclarationAnnotation is an abstract base, need to call annotation on the concrete class first."
-	);
-	return dynamic_cast<CallableDeclarationAnnotation&>(*m_annotation);
+	return abstractAnnotation<CallableDeclarationAnnotation>("CallableDeclarationAnnotation");
```

Comment:
Ah I see this one is different.

Reply:
Wondering if the new template function is necessary since it is used only once. Could it be used more frequently in the future? Couldn't we do:

```
solAssert(
	m_annotation,
	"CallableDeclarationAnnotation is an abstract base, need to call annotation on the concrete class first."
);
CallableDeclarationAnnotation& cda = initAnnotation<CallableDeclarationAnnotation>();
```

`m_annotation` might need to be moved back to the protected section though.

Reply:
Actually, since `CallableDeclaration` can never be the final overrider, would it make sense to have `CallableDeclaration::annotation` be pure virtual? I've checked and it compiles, and I see no reason it wouldn't work, since neither the constructor nor destructor calls `annotation`.

Reply:
Yes, sounds good!